### DISCUSSION TITLE
Does this program working with virtual desktop?

### DISCUSSION THREAD

**Discussion by kittinzaa** [kangyu-california/PersistentWindows#172] [Q&A] [answered]
I tend to have many virtual desktops opened so need to clearify this before I will try it. Thanks.

**Comment by kangyu-california** [comment]
PW is mainly designed for multi-monitor display (applicable for both local machine and remote desktop access).
It seems to work well with virtual desktops on Windows 10, but I have not tested this setup extensively.
Let me know your experience.

**Comment by jcbaillie** [comment]
The "restore windows from disk" feature does not work on multi desktops (at least not on my recent Windows 10): it will reopen all the windows on the current desktop, instead of reopening them on the desktop they belonged to when the "capture windows to disk" was done.

**Reply by kangyu-california** [reply]
I have added preliminary support for restoring window from disk to exact virtual desktop, 

To try the new enhancement, don't forget to redo "capture to disk", virtual desktop info will be captured in db file

Here is the patch to try [PersistentWindows5.42_p172.zip](https://github.com/kangyu-california/PersistentWindows/files/12063337/PersistentWindows5.42_p172.zip)

**Comment by kangyu-california** [comment]
The major obstacle to support VirtualDesktop-aware restore in PW is due to the lack of standard and stable C# API from Microsoft,

The burden is on individual developers side to choose a 3rd party wrapper package that works.
One of the 3rd party package is https://github.com/MScholtes/VirtualDesktop, which has 6 variations, each one is customized for a specific Windows OS version, this would be a nightmare for software maintenance.

**Reply by kangyu-california** [reply]
I have added preliminary support for restoring window from disk to exact virtual desktop, 

To try the new enhancement, don't forget to redo "capture to disk", virtual desktop info will be captured in db file

Here is the patch to try [PersistentWindows5.42_p172.zip](https://github.com/kangyu-california/PersistentWindows/files/12063337/PersistentWindows5.42_p172.zip)

**Reply by jcbaillie** [reply]
This is great, thank you! I cannot test it right now, I'm traveling, but I'll test and report here by end of July.

**Reply by kangyu-california** [reply, marked as answer]
Please try the latest patch instead [PersistentWindows5.43_p239_240_241.zip](https://github.com/kangyu-california/PersistentWindows/files/12139138/PersistentWindows5.43_p239_240_241.zip)

**Reply by jcbaillie** [reply]
Hi, I just tried it. It sort of works, but some windows are still misplaced. I think I have an idea what is going wrong: one of the program that is restarted is Brave, which I run with several different windows opened on different desktops. When Brave is restarted, it reopens all the windows in their respective desktop, which in itself is cool. But it must do its own thing at the same time that you do yours, so I would guess it pulls the rug under your feet when you think you are spawning a window in a particular desktop that you just activated, Brave is actually switching to another desktop without you knowing. A possible option would be to have a long delay when launching Brave (10s at least), to give it enough time to restart its own things, and then resume your work? NB: to make things worse, I actually have several Brave profiles which are spawned independently and which will all try to respawn their respective windows all around the place...

**Reply by kangyu-california** [reply]
Windows 10 can restore Brave/Chrome windows to the corresponding virtual desktop upon system restart without the help of PW. So user should not have this issue for daily usage.

Here is another patch with slightly improved dialog prompt upon switching virtual desktop. The key point of using a dialog is to let user make the decision as to when to switch virtual desktop and whether to launch a new window instead of mandatary 10s delay.

Lastly if not all windows are created and restored, user always has the option to do restore from disk again.

[PersistentWindows5.43_0726.zip](https://github.com/kangyu-california/PersistentWindows/files/12177145/PersistentWindows5.43_0726.zip)

**Reply by jcbaillie** [reply]
OK, I tried with manually validating each process restauration with "Yes" (instead of Yes to all), so it improves things a bit. I still have visual studio opening on the wrong desktop somehow. Also only one of my brave profile set of windows is relaunched (there are 3 profiles, so I have to relaunch two of them manually). Brave profiles are all launched with `brave.exe --profile-directory="ProfileXYZ"`, so they appear as one process I guess. Overall, still a huge time saver each time I have to reboot because Windows orders me to...
